python-checks/src/main/java/org/sonar/python/checks/tests/NotDiscoverableTestMethodCheck.java

```diff
@@ -0,0 +1,130 @@
+  // All methods of unittest https://docs.python.org/3/library/unittest.html#unittest.TestCase
+  private static final List<String> UNITTEST_RUN_METHODS = List.of("setUp", "tearDown", "setUpClass", "tearDownClass", "run", "skiptTest",
+    "subTest", "debug");
+  private static final List<String> UNITTEST_CHECK_METHODS = List.of("assertEqual", "assertNotEqual", "assertTrue", "assertFalse",
+    "assertIs", "assertIsNot", "assertIsNone", "assertIsNotNone", "assertIn", "assertNotIn", "assertIsInstance", "assertNotIsInstance",
+    "assertRaises", "assertRaisesRegex", "assertWarns", "assertWarnsRegex", "assertLogs", "assertNoLogs", "assertAlmostEqual", "assertGreater",
+    "assertGreaterEqual", "assertLess", "assertLessEqual", "assertRegex", "assertNotRegex", "assertCountEqual", "addTypeEqualityFunc",
+    "fail", "failureException", "longMessage", "maxDiff");
+  private static final List<String> UNITTEST_GATHER_INFO = List.of("countTestCases", "defaultTestResult", "id", "shortDescription", "addCleanup",
+    "doCleanups", "addClassCleanup", "doClassCleanups");
+  private static final List<String> UNITTEST_METHODS = concat(concat(UNITTEST_CHECK_METHODS.stream(), UNITTEST_RUN_METHODS.stream()), UNITTEST_GATHER_INFO.stream())
+    .collect(Collectors.toList());
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CLASSDEF, ctx -> {
+      ClassDef classDefinition = (ClassDef) ctx.syntaxNode();
+
+      if (inheritsOnlyFromUnitTest(classDefinition)) {
+        HashMap<String, List<Tree>> functionOccurrences = new HashMap<>();
+        List<FunctionDef> allDefinitions = new ArrayList<>();
+        // Assuming there is no definition of function in another one
+        for (Statement statement : classDefinition.body().statements()) {
+          if (statement.is(Tree.Kind.FUNCDEF)) {
+            FunctionDef functionDef = ((FunctionDef) statement);
+            String nameFunction = functionDef.name().name();
+            // If it doesn't override existing methods and doesn't start with test, we add it to the map
+            if (!overrideExistingMethod(nameFunction) && !nameFunction.startsWith("test")) {
+              addValueToMap(functionOccurrences, functionDef.name(), nameFunction);
+            }
+            allDefinitions.add(functionDef);
+          }
+        }
+
+        // Add tree everytime a token contains the name of function (call + reference functions are then taken care of)
+        allDefinitions.forEach(def -> retrieveReferenceInBody(def, functionOccurrences));
+
+        // More than 1 occurrence of the method means it is used inside the class -> do not raise
+        // Only raise when one (non-test) method is not used
+        functionOccurrences.entrySet()
+          .stream()
+          .filter((nameListEntry -> nameListEntry.getValue().size() == 1))
+          .forEach(nameListEntry -> ctx.addIssue(nameListEntry.getValue().get(0), MESSAGE));
+      }
```

Comment:
I think this part is a bit complex. What you're doing is going through each method definition's body to try to find usages of the suspicious methods, to check if they are actually helpers.

You can do this more simply by using the semantic API of sonar-python. In particular, when you have a `FunctionDef#name`, its associated symbol should most likely be a `FunctionSymbol` (if it's not the case, it may be because the symbol is defined multiple times, or other edge cases like that, for which we don't want to report issues anyway). 

When you have a `Symbol`, you can access its usages through the `Symbol#usages` API. By default, there is always at least one usage (the definition of the symbol). If you have more, you can check that the associated tree has indeed one of the method definitions as its parent, by doing something like:


```
suspiciousFunctionsAndDefinitions.forEach((s, d) -> {
  List<Usage> usages = s.usages();
  if (usages.size() == 1 || usages.stream().noneMatch(u -> TreeUtils.firstAncestor(u.tree(), allDefinitions::contains) != null)) {
    ctx.addIssue(d.name(), MESSAGE);
 }
});
```

Then you can remove `retrieveReferenceInBody` and `addValueToMap` which are now unused.